File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    static final int MAX = 500;
    int[][] dp;
    int[][] values;

    public int solution(int[][] triangle) {
        int answer = 0;

        dp = makeDP(triangle.length);
        values = triangle;

        answer = solve(0, 0);

        return answer;
    }

    int[][] makeDP(final int length) {
        int[][] results = new int[length][length];

        for (int y = 0; y < length; y++) {
            for (int x = 0; x < length; x++) {
                results[y][x] = -1;
            }
        }

        return results;
    }

    int solve(int height, int index) {
        if (height == values.length - 1) {
            return values[height][index];
        }

        if (dp[height][index] != -1) {
            return dp[height][index];
        }

        int present = values[height][index];

        dp[height][index] = Math.max(solve(height + 1, index), solve(height + 1, index + 1)) + present;

        return dp[height][index];
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

#define MAX 500

/*
    dp[i][j] : i번쨰 j 칸의 가능한 최댓값
    7
    3 8
    8 1 0
    2 7 4 4
    4 5 2 6 5
*/

int dp[MAX][MAX];

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    
    dp[0][0] = triangle[0][0];
    
    for (int i = 1; i < triangle.size(); i++) {
        for (int j = 0; j < triangle[i].size(); j++) {
            if (j == 0) {   // left
                dp[i][j] = dp[i - 1][j] + triangle[i][j];
            } else {
                int tempMax = max(dp[i - 1][j], dp[i - 1][j - 1]);
                dp[i][j] = tempMax + triangle[i][j];
            }
        }
    }
    
    int bottom = triangle.size() - 1;
    for (int x = 0; x < triangle[bottom].size(); x++) {
        answer = max(answer, dp[bottom][x]);
    }
    
    return answer;
}